File: SAW_bB.py
# import random
import random

# To determine the direction
# 1 for +x
# 2 for +y
# 3 for -x
# 4 for -y
direction = [1, 2, 3, 4]

# initial value assignment:
k = 0
generator = 0
rand_x = 0
rand_y = 0
row = 1000
col = 1000

# defining the 2D grid for Self Avoiding random walk (SAW):
# written in this way else one is unable to change a single element of the grid!
grid = [[0 for i in range(2 * col)] for j in range(2 * row)]

# inputs
print("Enter the total number of steps:")
steps = int(input())
print("Enter the step length:")
l = int(input())
print("Initial coordinates (x,y) one after another:")
x = int(input())
y = int(input())

rand_x = x + row
rand_y = y + col



grid[rand_x][rand_y] = 1  # initial position
# Finds single SAW path for given inputs:
while steps != 0:
    generator = random.choice(direction)

    if grid[rand_x + l][rand_y] == 0 and generator == 1:
        rand_x = rand_x + l
        grid[rand_x][rand_y] = 1  # final position
        steps = steps - 1
        print("+x", rand_x, rand_y)
    elif grid[rand_x][rand_y + l] == 0 and generator == 2:
        rand_y = rand_y + l
        grid[rand_x][rand_y] = 1  # final position
        steps = steps - 1
        print("+y", rand_x, rand_y)
    elif grid[rand_x - l][rand_y] == 0 and generator == 3:
        rand_x = rand_x - l
        grid[rand_x][rand_y] = 1  # final position
        steps = steps - 1
        print("-x", rand_x, rand_y)
    elif grid[rand_x][rand_y - l] == 0 and generator == 4:
        rand_y = rand_y - l
        grid[rand_x][rand_y] = 1  # final position
        steps = steps - 1
        print("-y", rand_x, rand_y)
    else:
        continue

File: SAW_bB.f
      integer steps,i,j,iter
      integer x,y, rand_x, rand_y,l
      real distance_sq, generator
      real distance_sq_sum, rms
      integer row, col, dx,dy
      integer grid(10000,10000) !defining the 2d lattice for walk
      integer c(10000),check_step

c Initializing values:
      rms=0.0
      distance_sq_sum=0.0
      distance_sq=0.0
      row=1000
      col=1000
      check_step=0

      do i=1, row
        do j=1, col
            grid(i,j)=0
        enddo
      enddo


c Taking input values:
      write(*,*)'Enter the total number of steps:'
      read(*,*) steps ! total #s of steps
      write(*,*) 'Enter the step length'
      read(*,*) l ! # step size
  !    write(*,*) 'Enter the number of iterations'
   !   read(*,*) iter !# of loops
      write(*,*) ' Initial coordinates (x,y) one after another:'
      read(*,*) x, y ! Initial coordinates assignment

c Copying initial coordinates to another variable for further analysis:
      rand_x=x+(row)/2
      rand_y=y+(col)/2


      grid(rand_x,rand_y)=1 !intial position

      do while(steps.ne.0)

10      generator= rand()  !direction
c====================================================================
        if(grid(rand_x+l,rand_y).eq.0 .and.generator.ge.0.0
     & .and. generator .le. 0.25) then
          rand_x=rand_x+l
         grid(rand_x,rand_y)=1
         steps=steps-1
        print*,"+x: Final coordinates (x,y):"
     &  ,rand_x-x-(row)/2,rand_y-y-(col)/2
         endif
c====================================================================
        if(grid(rand_x,rand_y+l).eq.0 .and.generator.gt.0.25
     & .and. generator .lt. 0.5 ) then
         rand_y=rand_y+l
         grid(rand_x,rand_y)=1
          steps=steps-1
      print*,"+y: Final coordinates (x,y):"
     &  ,rand_x-x-(row)/2,rand_y-y-(col)/2
         endif
c====================================================================
        if(grid(rand_x-l,rand_y).eq.0 .and. generator.gt.0.5
     & .and. generator .lt. 0.75 ) then
        rand_x=rand_x-l
        grid(rand_x,rand_y)=1
         steps=steps-1
         print*,"-x: Final coordinates (x,y):"
     &  ,rand_x-x-(row)/2,rand_y-y-(col)/2
         endif
c====================================================================

         if(grid(rand_x,rand_y-l).eq.0 .and. generator.gt.0.75
     & .and. generator .le. 1.0 ) then
         rand_y=rand_y-l
        grid(rand_x,rand_y)=1
         steps=steps-1
       print*,"-y: Final coordinates (x,y):"
     &  ,rand_x-x-(row)/2,rand_y-y-(col)/2
         endif
c====================================================================

         enddo



      stop
      end



